M2/Macaulay2/packages/FastLinAlg.m2

```diff
@@ -855,6 +940,71 @@ projDim(Module) := opts -> (N1) -> (
     return i;
 );
 
+isGBDone := (myGB) -> (
+    --a temporary function for finding out if a gb computation is done.
+    myStr := status myGB;
+    if (0 < #select("status: done", myStr)) then return true else return false;
+);
+
+
+
+isCodimAtLeast = method(Options => optIsCodimAtLeast);
+
+isCodimAtLeast(ZZ, Ideal) := opts -> (n1, I1) -> (
+    R1 := ring I1;
+    S1 := ambient R1;
+    if (not isPolynomialRing(S1)) then error "isCodimAtLeast:  This requires an ideal in a polynomial ring, or in a quotient of a polynomial ring.";
+    if (n1 <= 0) then return true; --if for some reason we are checking codim 0.
+    if (isMonomialIdeal(I1)) then (
+        if (codim monomialIdeal(I1) >= n1) then return true;
+    );
+    if (#first entries gens I1 == 0) then ( return false; ); 
+    J1 := ideal R1;
+    dAmb := codim J1;
+    I2 := sub(I1, S1) + sub(J1, S1); --lift to the polynomial ring.
+    --now we have an ideal in a polynomial ring.  The idea is that we should compute a partial Groebner basis.
+    --and compute the codim of that.
+    --But first, we just try a quick codim computation based on the ideal generators.
+    monIdeal := null;
+    if (#first entries gens I2 > 0) then (
+        monIdeal = monomialIdeal(apply(first entries gens I2, t->leadTerm t));
+        if (opts.Verbose or debugLevel > 0) then print concatenate("isCodimAtLeast: Computing codim of monomials based on ideal generators.");
+        if (codim monIdeal - dAmb >= n1) then return true;
+        if (opts.Verbose or debugLevel > 0) then print concatenate("isCodimAtLeast: Didn't work, going to find the partial Groebner basis.");
+    );
+    --now we set stuff up for the loop if that doesn't work.  
+    vCount := # first entries vars S1;
+ --   baseDeg := apply(sum(apply(first entries vars S1, t1 -> degree t1)), v -> ceiling(v/vCount)); --use this as the base degree to step by (probably we should use a different value)
+    i := 1;
+    --curLimit := baseDeg;
+    local curLimit;
+    local myGB;
+    
+    gensList := null;
+    while (i < opts.PairLimit) do(
+        curLimit = opts.SPairsFunction(i);
+--        curLimit = apply(baseDeg, tt -> (opts.SPairsFunction)(tt,i));
+        if (opts.Verbose or debugLevel > 0) then print concatenate("isCodimAtLeast: about to compute gb PairLimit => ", toString curLimit);
+        myGB = gb(I2, PairLimit=>curLimit);
+        gensList = first entries leadTerm gb(I2, PairLimit=>curLimit);    
+        if (#gensList > 0) then (
+            monIdeal = monomialIdeal(first entries leadTerm myGB);
+            if (opts.Verbose or debugLevel > 0) then print concatenate("isCodimAtLeast: computed gb, now computing codim ");
+            if (codim monIdeal - dAmb >= n1) then return true;
+        );
+        if (isGBDone(myGB)) then i = opts.PairLimit;
+        i = i + 1;
+    );
+    return null;
```

Comment:
This line can be omitted.

Reply:
I realize that, but my authors and I thought this way would be easeir to read.